### PR TITLE
Hide user email partially

### PR DESCRIPTION
**Pull request by Rup-Narayan-Rajbanshi**
Addresses User

## Changes
- Add a function to hide email of a user

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [x] permission checks (tests here too)
- [x] translations

### CODE REVIEW

File: apps/user/schema.py [hunk 1]
```diff
@@ -47,6 +48,7 @@ class Meta:
     display_picture_url = graphene.String()
     organization = graphene.String()
     language = graphene.String()
+    email_display = graphene.String()
```
Comment (thenav56) [inline apps/user/schema.py]:
```suggestion
    email_display = graphene.String(required=True)
```

File: apps/user/schema.py [hunk 2]
```diff
@@ -58,6 +60,11 @@ def resolve_display_picture_url(root, info, **kwargs) -> Union[str, None]:
     def resolve_organization(root, info, **kwargs) -> Union[str, None]:
         return info.context.dl.user.organization.load(root.id)
 
+    @staticmethod
+    def resolve_email_display(root, info, **kwargs) -> Union[str, None]:
+        email_display = generate_hidden_email(root.email)
+        return email_display
```
Comment (thenav56) [inline apps/user/schema.py]:
```suggestion
        return generate_hidden_email(root.email)
```

File: apps/user/utils.py [hunk 3]
```diff
@@ -197,3 +197,20 @@ def send_password_changed_notification(user_id, client_ip, device_type):
         subject_template_name='password_changed/subject.txt',
         email_template_name='password_changed/email.html',
     )
+
+
+def generate_hidden_email(email):
```
Comment (thenav56) [inline apps/user/utils.py:202]: We need an in-depth test for this function.

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/user/schema.py [hunk 4]
```diff
@@ -47,6 +48,7 @@ class Meta:
     display_picture_url = graphene.String()
     organization = graphene.String()
     language = graphene.String()
+    email_display = graphene.String(required=True)
```
Comment (AdityaKhatri) [inline apps/user/schema.py:51]: Change this to display_email.
Right @thenav56 ?

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: apps/user/schema.py (hunk 4 above)
Comment (thenav56) [reply apps/user/schema.py:51]: `email_display` is fine.

File: apps/user/tests/test_schemas.py [hunk 5]
```diff
@@ -563,3 +564,88 @@ def _query_check(filters, **kwargs):
...
+        user4 = UserFactory.create(email='abc@deep.com')
+        user5 = UserFactory.create(email='xy@deep.com')
+        user6 = UserFactory.create(email='a@deep.com')
+
+        user1_expected_hidden_email = 'te...r1@deep.com'
+        user2_expected_hidden_email = 'te...r2@deep.com'
+        user3_expected_hidden_email = 'a...d@deep.com'
+        user4_expected_hidden_email = 'a...c@deep.com'
+        user5_expected_hidden_email = 'x...y@deep.com'
+        user6_expected_hidden_email = 'a...a@deep.com'
```
Comment (thenav56) [inline apps/user/tests/test_schemas.py]: @AdityaKhatri Are these cases enough?

File: apps/user/tests/test_schemas.py [hunk 6]
```diff
@@ -563,3 +564,88 @@ def _query_check(filters, **kwargs):
...
+        user4_generated_email = generate_hidden_email(user4.email)
+        user5_generated_email = generate_hidden_email(user5.email)
+        user6_generated_email = generate_hidden_email(user6.email)
+
+        self.assertEqual(user1_expected_hidden_email, user1_generated_email)
+        self.assertEqual(user2_expected_hidden_email, user2_generated_email)
+        self.assertEqual(user3_expected_hidden_email, user3_generated_email)
+        self.assertEqual(user4_expected_hidden_email, user4_generated_email)
+        self.assertEqual(user5_expected_hidden_email, user5_generated_email)
+        self.assertEqual(user6_expected_hidden_email, user6_generated_email)
```
Comment (thenav56) [inline apps/user/tests/test_schemas.py]: @Rup-Narayan-Rajbanshi Let's only test the function, UserFactory is not required here.

```suggestion
        for original, expected in [
            ('testuser1@deep.com', 'te...r1@deep.com'),
            ('testuser2@deep.com', 'te...r2@deep.com'),
            ('abcd@deep.com', 'a...d@deep.com'),
            ('abc@deep.com', 'a...c@deep.com'
            ('xy@deep.com', 'x...y@deep.com'),
            ('a@deep.com', 'a...a@deep.com'),
        ]:
            self.assertEqual(expected, generate_hidden_email(original))
```

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: apps/user/tests/test_schemas.py (hunk 5 above)
Comment (AdityaKhatri) [reply apps/user/tests/test_schemas.py]: This should be good.
@Rup-Narayan-Rajbanshi , can you change the `.` by `*`?

Comment (Rup-Narayan-Rajbanshi) [reply apps/user/tests/test_schemas.py]: ok @AdityaKhatri dai.

**Review by thenav56 — APPROVED** [review]

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**